### PR TITLE
Adding KNN scorer explanation

### PR DESCRIPTION
**Pull request by metonymic-smokey**
This PR adds a score Explanation to the KNN scorer, along with a unit test.
Also, contains a minor refactor of scoring related code.

### CODE REVIEW

**Review by abhinavdangeti — APPROVED** [review]
Looks fine to me.
@metonymic-smokey would you add a sample snippet of the "explain score" output to the commit message here.

Also, I'm glad you've set the base branch to `MB-59102` so there'd be no conflicts, but I think we should merge this into unstable after the base PR has been so we don't squash this into it and create a separate commit entry for this.

**Comment by metonymic-smokey** [comment]
Done, pls let me know what you think.
thanks!

File: search/scorer/scorer_knn.go [hunk 1]
```diff
@@ -73,9 +71,36 @@ func (sqs *KNNQueryScorer) Score(ctx *search.SearchContext,
...
 			score = score * sqs.queryWeight
+			if sqs.options.Explain {
+				childExplanations := make([]*search.Explanation, 2)
+				childExplanations[0] = sqs.queryWeightExplanation
+				childExplanations[1] = scoreExplanation
+				scoreExplanation = &search.Explanation{
+					Value: score,
+					// Product of score * weight
+					Message: fmt.Sprintf("weight(%s:%f^%f in %s), product of:",
+						sqs.queryField, sqs.queryVector, sqs.queryBoost, knnMatch.ID),
```
Comment (abhinavdangeti) [inline search/scorer/scorer_knn.go:100]: Just noticed this, let's not emit the actual query vector to the output - there'll be an overload of data.

File: search/scorer/scorer_knn.go [hunk 2]
```diff
@@ -100,4 +125,22 @@ func (sqs *KNNQueryScorer) SetQueryNorm(qnorm float64) {
...
+			Message: "boost",
+		}
+		childrenExplanations[1] = &search.Explanation{
+			Value:   sqs.queryNorm,
+			Message: "queryNorm",
+		}
+		sqs.queryWeightExplanation = &search.Explanation{
+			Value: sqs.queryWeight,
+			Message: fmt.Sprintf("queryWeight(%s:%f^%f), product of:",
+				sqs.queryField, sqs.queryVector, sqs.queryBoost),
```
Comment (abhinavdangeti) [inline search/scorer/scorer_knn.go:142]: Ditto.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Comment by metonymic-smokey** [comment]
Pls review this PR instead - https://github.com/blevesearch/bleve/pull/1899

**Pull request closed without merging**